packages-next/fields-document/src/component-blocks.tsx

```diff
@@ -0,0 +1,206 @@
+            <Checkbox
+              checked={value}
+              onChange={event => {
+                onChange(event.target.checked);
+              }}
+            >
+              {label}
+            </Checkbox>
+          </FieldContainer>
+        );
+      },
+      defaultValue: false,
+    };
+  },
+  empty(): FormField<undefined> {
+    return {
+      kind: 'form',
+      Input() {
+        return null;
+      },
+      defaultValue: undefined,
+    };
+  },
+  child(): InlineField {
+    return {
+      kind: 'inline',
+    };
+  },
+  object<Value extends Record<string, ComponentPropField>>(value: Value): ObjectField<Value> {
+    return { kind: 'object', value };
+  },
+  conditional<
+    Discriminant extends string | boolean,
+    ConditionalValue extends [Discriminant] extends [boolean]
+      ? { true: ComponentPropField; false: ComponentPropField }
+      : [Discriminant] extends [string]
+      ? { [Key in Discriminant]: ComponentPropField }
+      : never
+  >(
+    discriminant: FormField<Discriminant>,
+    values: ConditionalValue
+  ): ConditionalField<Discriminant, ConditionalValue> {
+    return {
+      kind: 'conditional',
+      discriminant,
+      values: values,
+    };
+  },
+};
+
+export type ComponentBlock = {
+  component: (props: any) => ReactElement | null;
+  props: Record<string, ComponentPropField>;
+};
+
+type DiscriminantToString<Discriminant extends string | boolean> = Discriminant extends boolean
+  ? 'true' | 'false'
+  : Discriminant;
+
+type CastToComponentPropField<Prop> = Prop extends ComponentPropField ? Prop : never;
+
+type ExtractPropFromComponentPropField<Prop extends ComponentPropField> = [Prop] extends [
+  InlineField
+]
+  ? ReactNode
+  : [Prop] extends [FormField<infer Value>]
+  ? Value
+  : [Prop] extends [ObjectField<infer Value>]
+  ? { [Key in keyof Value]: ExtractPropFromComponentPropField<Value[Key]> }
+  : [Prop] extends [ConditionalField<infer Discriminant, infer Value>]
+  ? {
+      [Key in DiscriminantToString<Discriminant>]: {
+        discriminant: Discriminant extends boolean
+          ? 'true' extends Key
+            ? true
+            : 'false' extends Key
+            ? false
+            : never
+          : Discriminant;
+        value: Key extends keyof Value
+          ? ExtractPropFromComponentPropField<CastToComponentPropField<Value[Key]>>
+          : never;
+      };
+    }[DiscriminantToString<Discriminant>]
+  : 'fields must be one of ComponentPropField, not a union of multiple of them';
+
+export function component<
+  PropsOption extends {
+    [Key in any]: ComponentPropField;
+  }
+>(options: {
+  component: (
+    props: {
+      [Key in keyof PropsOption]: ExtractPropFromComponentPropField<PropsOption[Key]>;
+    }
+  ) => ReactElement | null;
+  props: PropsOption;
+}): ComponentBlock {
+  return options;
+}
+
+export const NotEditable = ({ children, ...props }: { children: ReactNode }) => (
+  <span contentEditable={false} {...props}>
```

Comment:
Why doesn't this have `css={{ userSelect: 'none' }}` on it? I thought that was also important for non-editable things in Slate

Reply:
Added it back. I was testing things to see when it fails without user-select: none (if you know how to get it to fail, I would love to know)